nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,138 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(%{shape: {_m, k} = shape_q} = q, %{shape: {k, n}} = r, opts) do
+    mode = opts[:mode]
+
+    keep_reflectors = mode == :raw
+    calculate_q = mode in [:reduced, :complete, :raw]
+
+    q =
+      if calculate_q do
+        identity(q, shape: shape_q)
+      end
```

Comment:
As @elbow-jason mentioned it's expensive to build this and reduce on the pre-built result. I'm not familiar with the algorithm, but if possible you should try and iteratively build the binary by appending to it.

Reply:
This algorithm relies heavily on pre and post matrix multiplication to yield Q and R. I'll try to think of a more iterative way so we can build the binary dinamically.